OpenSim/Common/Exception.h

```diff
@@ -88,11 +127,42 @@ class OSIMCOMMON_API Exception  : public std::exception {
     virtual void print(std::ostream &aOut) const;
 #endif
     // override virtual function from std::exception
-    const char* what() const throw() override {return getMessage();}
+    const char* what() const noexcept override;
 
 //=============================================================================
 };  // END CLASS Exception
 
+
+class IndexOutOfRange : public Exception {
+public:
+    IndexOutOfRange(const std::string& file,
+                    size_t line,
+                    const std::string& func,
+                    size_t index, 
+                    size_t min, 
+                    size_t max) :
+        Exception(file, line, func) {
+        std::string msg = "min = " + std::to_string(min);
+        msg += " max = " + std::to_string(max);
+        msg += " index = " + std::to_string(index);
+
+        addMessage(msg);
```

Comment:
I think this is fine. To do it in the base class constructor call would require writing an additional method in each exception subclass, and that method would duplicate the signature of the subclass constructor except with `file,line,func` removed. That doesn't look like much of a burden here, where it would be something like `Exception(file,line,func,formatMsg(index,min,max))`, but consider a typical Simbody exception class:
```c++
class ErrorCheck : public Base {
public:
    ErrorCheck(const char* fn, int ln, const char* assertion, 
           const char* whereChecked,    // e.g., ClassName::methodName()
           const char* fmt ...) : Base(fn,ln)
    {
        char buf[1024];
        va_list args;
        va_start(args, fmt);
        vsprintf(buf, fmt, args);

        setMessage("Error detected by Simbody method " 
            + std::string(whereChecked) + ": "
            + std::string(buf)
            + "\n  (Required condition '" + std::string(assertion) + "' was not met.)\n");
        va_end(args);
    }
};
```
It would be no fun to have to forward that varargs constructor to a private method that formats the message! This might be a little easier use C++11 varargs capability, but it still seems like a lot of work with potential bugs produced in rarely-tested exception message generation.

One discipline we could follow would be to always format the message directly into the `addMessage()` argument so that it would be impossible to forget to call it.